02_calculator/calculator.rb

```diff
@@ -1 +1,28 @@
-#write your code here
+
+
+def add ( num1,num2)
+
+   
+  add=num1 + num2
+end
+add(0,0)
+
+
+def subtract ( num1,num2)
+
+   
+  subtract=num1 - num2
+    end
+subtract(10,4)
+
+    
+    
+  
+     
+def sum array
+  sum = 0
+  array.map! do |number| 
```

Comment:
`map!` overwrites the original array you passed in, which does more than you want and could lead to nasty surprises. For example, here's what happens when I copy your function into `irb` and use it:

![Screenshot 2023-04-14 at 7 10 23 PM](https://user-images.githubusercontent.com/5075438/232169013-42730386-53cf-47f7-b6d4-ca222171cf6c.png)

Notice that it reads `[1, 3, 6]` on the last line, even though I defined `my_array` as `[1, 2, 3]`, and the sum of `[1, 3, 6]` is actually 10. Can you figure out where the 3 and 6 came from?

A less destructive method you could use is `each`:

```suggestion
  array.each do |number| 
```

Reply:
Oh I see. I debug through rake. but I tested my code in treehouse workspace and I got the same. I see how treehouse work space lets me know where the errors. I did some research about using .map! verses using .map. It says it mutates the original array . I'm assuming it mutated through each number that's how  [1,2,3] became [1,3,6]. However, when I delete the exclamation mark it says "6 is the sum of [1,2,3]

Reply:
`.map` so happens to work because it doesn't mutate the original array; however, it's still doing more than you need compared to `each` because `map` creates a separate array based on the value returned by the block, which is a waste of memory when all you need is `sum`. For example, if I were to assign the result of the map to a variable and print it like so:

```ruby
def sum array
  sum = 0
  map_result = array.map do |number| 
    sum+= number
  end
  puts map_result.inspect
  sum
end

sum([1, 2, 3])
```

Then we see the same `[1, 3, 6]` earlier in the output.

> I'm assuming it mutated through each number that's how [1,2,3] became [1,3,6].

Specifically, the end of your block is the line `sum+= number`, so after Ruby adds the current number to `sum`, it returns the value of `sum`. That means the array `[1, 3, 6]` is the [running total](https://en.wikipedia.org/wiki/Running_total) at each index.

1 = **1**
3 = 1 + **2**
6 = 1 + 2 + **3**